src/models/User.js

```diff
@@ -0,0 +1,50 @@
+const bcrypt = require("bcrypt");
+const mongoose = require("mongoose");
+const validator = require("validator");
+
+const userSchema = mongoose.Schema(
+  {
+    name: {
+      type: String,
+      required: [true, "please provide a name!"],
+    },
+    email: {
+      type: String,
+      required: [true, "please provide an email!"],
+      validate: {
+        validator: validator.isEmail,
+        message: "Please provide a valid email",
+      },
+      unique: true,
+    },
+    password: {
+      type: String,
+      required: [true, "please provide a password!"],
+      minlength: [8, "password must be at least 8 characters long!"],
+      select: false,
+    },
+    groups: [{ type: mongoose.Schema.Types.ObjectId, ref: "Group" }],
+
+    // maybe not for the MVP
+    // role: {
+    //   type: String,
+    //   enum: {
+    //     values: ["user", "admin"],
+    //     message: "Invalid role!",
+    //   },
+    //   required: [true, "please provide a valid role!"],
+    // },
+  },
+  { timestamps: true }
+);
+
+userSchema.pre("save", async function () {
+  const salt = await bcrypt.genSalt();
+  this.password = await bcrypt.hash(this.password, salt);
+});
+
+userSchema.methods.comparePassword = async function (inputtedPassword) {
+  return await bcrypt.compare(inputtedPassword, this.password);
+};
+
+module.exports = mongoose.model("User", userSchema);
```

Comment:
Referencing your question in slack -- please do feel free to create tasks. It's best to only include the work required to satisfy a ticket in the associated PR, and if there's anything extra, you'll usually want to call it out and explain why it should be included. This is so that we keep deploys clean and isolated and we know what to test (because it's explicitly called out in the ticket). Often in a real-world app, if there's a big problem with some code that's been deployed, it might be rolled back to get it out of production as quickly as possible. With this PR, we might add a problem with authentication, but it would result in us losing unrelated API functionality if we roll it back. Does that make sense?

So my suggestion is to also separate the auth code and create a ticket for that as well if possible, but I'd also be okay with just letting this be as is, and we can just keep things more isolated going forward. What do you think?

Reply:
Yeah this model shouldn't have been included into this PR. Will take it out of here and put it into the auth branch.